src/core/layertree/qgslayertreemodellegendnode.cpp

```diff
@@ -582,31 +601,83 @@ void QgsSymbolLegendNode::updateLabel()
 
   bool showFeatureCount = mLayerNode->customProperty( QStringLiteral( "showFeatureCount" ), 0 ).toBool();
   QgsVectorLayer *vl = qobject_cast<QgsVectorLayer *>( mLayerNode->layer() );
+  mLabel = symbolLabel();
 
-  if ( mEmbeddedInParent )
+  if ( showFeatureCount && vl )
   {
-    QString layerName = mLayerNode->name();
-    if ( !mLayerNode->customProperty( QStringLiteral( "legend/title-label" ) ).isNull() )
-      layerName = mLayerNode->customProperty( QStringLiteral( "legend/title-label" ) ).toString();
+    qlonglong count = mEmbeddedInParent ? vl->featureCount() : vl->featureCount( mItem.ruleKey() ) ;
+    mLabel += QStringLiteral( " [%1]" ).arg( count != -1 ? QLocale().toString( count ) : tr( "N/A" ) );
+  }
+
+  emit dataChanged();
+}
+
+QString QgsSymbolLegendNode::evaluateLabel( QgsExpressionContext context, const QString label )
+{
+  if ( !mLayerNode )
+    return QString();
 
-    mLabel = mUserLabel.isEmpty() ? layerName : mUserLabel;
-    if ( showFeatureCount && vl && vl->featureCount() >= 0 )
-      mLabel += QStringLiteral( " [%1]" ).arg( vl->featureCount() );
+  QgsVectorLayer *vl = qobject_cast<QgsVectorLayer *>( mLayerNode->layer() );
+
+  if ( label.isEmpty() )
+  {
+    if ( vl )
+    {
+      const QString symLabel = symbolLabel();
+      if ( ! mLayerNode->labelExpression().isEmpty() )
+        mLabel = evaluateLabelExpression( "[%" + mLayerNode->labelExpression() + "%]", context );
+      else if ( mLabel.contains( "[%" ) )
+        mLabel = evaluateLabelExpression( symLabel, context );
+    }
+    return mLabel;
   }
   else
   {
-    mLabel = mUserLabel.isEmpty() ? mItem.label() : mUserLabel;
-    if ( showFeatureCount && vl )
+    QString eLabel;
+    if ( vl )
     {
-      qlonglong count = vl->featureCount( mItem.ruleKey() );
-      mLabel += QStringLiteral( " [%1]" ).arg( count != -1 ? QLocale().toString( count ) : tr( "N/A" ) );
+      if ( ! mLayerNode->labelExpression().isEmpty() )
+        eLabel = evaluateLabelExpression( label + "[%" + mLayerNode->labelExpression() + "%]", context );
+      else if ( label.contains( "[%" ) )
+        eLabel = evaluateLabelExpression( label, context );
     }
+    return eLabel;
   }
-
-  emit dataChanged();
 }
 
+QgsExpressionContext QgsSymbolLegendNode::createExpressionContext( QgsExpressionContext context ) const
```

Comment:
This should better produce a "Symbol scope" only instead of a full context.

`std::unique_ptr<QgsExpressionContextScope> createSymbolScope() const;`